Feature/dd tools

EvgenKor opened the pull request:
The changes introduced by that PR are intended to support direct deposits

- Implemented a universal method `directDeposit` to deposit funds to the current account via direct deposit scheme (both token and native scheme are supported through the `type` parameter: `DirectDepositType.Token`\\`DirectDepositType.Native`). 
```
public async directDeposit(
    type: DirectDepositType, // DirectDepositType.Token or DirectDepositType.Native
    fromAddress: string,  // sender address to check approval and using as fallback address
    amount: bigint, // in pool resolution
    sendTxCallback: (tx: PreparedTransaction) => Promise<string>, // should returns txHash
  ): Promise<void>;
```
`sendTxCallback` will invoked to send transaction and `PreparedTransaction` is defined as following:
```
export interface PreparedTransaction {
    to: string; // dd queue contract address
    amount: bigint; // nonzero for native direct deposits
    data: string; // 0x-based hex string of encoded transaction
}
```
- introduced new optional parameters in Pool configuration: `isNative` (`boolean`, should be set as `true` for WETH-pools to accept native direct deposits) and `ddSubgraph` (`string`, the subgraph name to fetching pending direct deposits, currently supported `zkbob-bob-goerli`, `zkbob-eth-goerli` and `zkbob-bob-goerli-opt`)
- added method `getPendingDDs` to fetch pending direct deposits in the queue (available just for pools with configured subgraphs). It returns an array of `DirectDeposit` objects:
```
interface DirectDeposit {
    id: bigint;           // DD queue unique identifier
    state: DirectDepositState; // getPendingDDs will always return records
                               // with `DirectDepositState.Queued` state
    amount: bigint;       // in pool resolution
    destination: string;  // zk-addresss
    fallback: string;     // 0x-address to refund DD
    timestamp: number;    // when it was created
    queueTxHash: string;  // transaction hash to the queue
}
```
- added method `directDepositFee` which returns fee for the direct deposit in pool resolution (as returned by queue contract)
- added method `directDepositContract` to retrieve DD queue contract address from the pool
- added method `tokenSellerContract` to retrieve token seller contract address from the pool (also available in account-lessmode)

This PR can be tested with [associated console branch](https://github.com/zkBob/zkbob-console/pull/87). Pending direct deposits are currently fetched in `BOB-goerli`, `WETH-goerli` and `BOB-op-goerli` pool

Please note `BOB-goerli` pool is currently migrated on USDC and require associated changes in denominator processing, implemented in #147, so testing that branch on `BOB-goerli` pool is undesirable

akolotov commented on src/dd.ts:
```diff
@@ -0,0 +1,156 @@
+
+    public async pendingDirectDeposits(): Promise<DirectDeposit[]> {
+        if (this.subgraphName == 'zkbob-bob-goerli') {
+            const allPendingDDs = (await execute(ddQuery, {})).data.directDeposits;
```
It makes sense to check sync state of the subgraph through  https://api.thegraph.com/index-node/graphql by:
```
{
  indexingStatusForCurrentVersion(subgraphName: "zkbob/zkbob-bob-goerli") {
    synced
    health
    fatalError {
      message
      block {
        number
        hash
      }
      handler
    }
    chains {
      chainHeadBlock {
        number
      }
      latestBlock {
        number
      }
    }
  }
}
```
Then `latestBlock` can be used to call `eth_getLogs` for `SubmitDirectDeposit`, `CompleteDirectDepositBatch`, `RefundDirectDeposit` events to identify pending deposits which were not handled by subgraph if it is stale.

EvgenKor replied:
I'm going to implement that. But currently I think we should publish the lib asap since we want to deploy weth pool this week. Gathering contract logs is a feature which should work as a subgraph in some sense. Maybe it's better to prepare such improvements a bit later?

akolotov replied:
Yes, it is ok to implement it later if we have the corresponding work item (e.g. a github issue) in the backlog.

EvgenKor replied:
https://github.com/zkBob/zkbob-client-js/issues/146

k1rill-fedoseev commented on src/networks/evm.ts:
```diff
@@ -161,25 +163,76 @@ export class EvmNetwork implements NetworkBackend {
+            this.poolContract().options.address = poolAddress;
+            tokenSellerAddr = await this.poolContract().methods.tokenSeller().call();
+            if (tokenSellerAddr) {
+                this.ddContractAddresses.set(poolAddress, tokenSellerAddr);
```
```suggestion
                this.tokenSellerAddresses.set(poolAddress, tokenSellerAddr);
```

EvgenKor replied:
Sure, thanks!

k1rill-fedoseev commented on src/dd.ts:
```diff
@@ -0,0 +1,156 @@
+                            timestamp: Number(dd.tsInit),
+                            queueTxHash: dd.txInit,
+                        };
+                        (await result).push(ownDD);
```
Why do we need await here?
```suggestion
                        result.push(ownDD);
```

EvgenKor replied:
It's because the result is a `Promise` due to reduce callback function is asynchronous.  Invoking of `push` method on `Promise` will cause an exception
<img width="779" alt="image" src="https://github.com/zkBob/zkbob-client-js/assets/1415489/5727e86e-76ca-4c9d-9bb7-2339eec511ca">

k1rill-fedoseev replied:
`result` will be an array during the first iteration, but a `Promise` in later iterations. I personally don't like an idea of creating a recursive promise here. It works, but it looks really strange to me, I think it's better if we use a more common `Promise.all` + `map` + `filter` approach.

EvgenKor replied:
Fixed in https://github.com/zkBob/zkbob-client-js/pull/144/commits/fe16410083cfe48924cb15b6c6201a3902ffc019

EvgenKor commented:
The latest changes:
- added support for the direct deposit limits (the data fetched from the contract directly until relayer doesn't support it)
- added ability to use subgraph with arbitrary name. Currently deployed and tested: `zkbob-bob-goerli`, `zkbob-eth-goerli` and `zkbob-bob-goerli-opt`

k1rill-fedoseev commented on src/config.ts:
```diff
@@ -32,10 +32,12 @@ export interface Pool {
   feeDecimals?: number;
-  depositScheme: DepositType;
+  isNative?: boolean;
+  ddSubgraph?: string;
```
Can we support a full graphQL url here?
We may want to host some of them in different places, like Chainstack or decentralized The Graph network, not only in their hosted version.

EvgenKor replied:
I think most likely we can. Will try to tune

EvgenKor replied:
Added universal endpoint support in https://github.com/zkBob/zkbob-client-js/pull/144/commits/65458abdf5e3a261217ee7f1978fb3c0591011ca Now it accepts both subgraph name in the our directory on the hosted service and full endpoint URL. The input string considered to be a full URL in case of character `/` existence

EvgenKor commented on src/client.ts:
```diff
@@ -704,6 +714,42 @@ export class ZkBobClient extends ZkBobProvider {
+      // For the token-based DD we should check allowance first
+      const curAllowance = await this.network().allowance(pool.tokenAddress, fromAddress, ddQueueAddress);
+      if (curAllowance < fullAmountNative) {
+        throw new TxDepositAllowanceTooLow(fullAmountNative, curAllowance, ddQueueAddress);
```
That code is the deposit conditions reserve checking since we decide the token approval approving stage is on the UI side. Also I think it isn't so good to implement **_just a part_** of approving stage without RPC endpoints sync

AllFi commented on src/dd/index.ts:
```diff
@@ -0,0 +1,164 @@
+                throw new InternalError(`Unexpected response from the DD subgraph: ${allPendingDDs}`);
+            }
+        } else {
+            console.warn('There is no configured subraph to query pending DD')
```
I think it makes sense to throw some error here so the caller can distinguish the case when the subgraph is not configured from the case when there are no pending deposits. But it is not very important so it's just a suggestion.

EvgenKor replied:
One of the following improvements will be direct DD requesting from the contract even in case of subgraph omit. So I believe UI shouldn't process such error now

AllFi replied:
Got it. I agree, it doesn't seem necessary in that case.

AllFi reviewed (approved):
Looks good to me
